design/external-authorization-design.md

```diff
@@ -0,0 +1,670 @@
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
+	// Namespace of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/
+	// +optional
+	Namespace string `json:"namespace,omitempty" protobuf:"bytes,2,opt,name=namespace"`
+	// Name of the referent.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	// +required
+	Name string `json:"name,omitempty" protobuf:"bytes,3,opt,name=name"`
+}
+
+type AuthorizationPolicy struct {
+	// When true, this field disables client request authentication
+	// for the scope of the policy.
+	//
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merged
+	// such that the inner scope overrides matching keys from the
+	// outer scope.
+	//
+	// +optional
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
+	// +required
+	ServiceRef SupportServiceReference `json:"serviceRef"`
+
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+
+	// Timeout configures maximum time to wait for a check response from the authorization server.
+	// Timeout durations are expressed in the Go [Duration format](https://godoc.org/time#ParseDuration).
+	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
+	// The string "infinity" is also a valid input and specifies no timeout.
+	//
+	// +optional
+	Timeout string `json:"timeout,omitempty"`
+
+    // If FailOpen is true, the client request is forwarded to the upstream service
+    // even if the authorization server fails to respond. This field should not be
+    // set in most cases. It is intended for use only while migrating applications
+    // from internal authorization to Contour external authorization.
+    //
+    // +optional
+    FailOpen bool `json:"failOpen,omitempty"`
+}
+
+// VirtualHost appears at most once. If it is present, the object is considered
+// to be a "root".
+type VirtualHost struct {
+    ...
+
+	// This field configures a support service to perform
+	// authorization for this virtual host. Authorization can
+	// only be configured on virtual hosts that have TLS enabled.
+	// If the TLS configuration requires client certificate
+	///validation, the client certificate is always included in the
+	// authentication check request.
+	//
+	// +optional
+	Authorization *AuthorizationServer `json:"authorization,omitempty"`
+}
+
+// Route contains the set of routes for a virtual host.
+type Route struct {
+    ...
+
+    // AuthPolicy updates the authorization policy for client
+	// requests that match this route.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```

Comment:
Is there a way that operators can not allow users to change this? I'm thinking that an operator wants to have auth on for a set of routes, what stops a user from circumventing that and disabling?

Reply:
You'd have to use an admission controller (gatekeeper or suchlike) to control this. Generally, I expect that the app owner is responsible for the app and best placed to tweak authorization.

Reply:
So this is the same exact use-case of `permitInsecure`. @youngnick I know you had thoughts around this that you don't like the arg today to disable this feature, but I think it needs thought out. 

If admission controller is the answer, we need to scope that work since many folks will want to enforce this from the start.

Reply:
> So this is the same exact use-case of `permitInsecure`. @youngnick I know you had thoughts around this that you don't like the arg today to disable this feature, but I think it needs thought out.
> 
> If admission controller is the answer, we need to scope that work since many folks will want to enforce this from the start.

I'm happy to have this discussion, but I don't think it's practical to have separate admin control over every field that can be specified in a HTTPProxy resource. General purpose admission controllers exist for this reason, and it's hard for me to imagine that doing the equivalent in Contour is really worth it.

Reply:
To be clear, @stevesloka, I don't like that we have to disable a permit globally, that's all. I agree that sometimes, cluster admins are going to want to say "you must have auth on, you can't turn it off", but that conflicts with the goal of being able to selectively disable some routes, which is more useful to start with, I think.

To put this another way, I think that being able to expose *some* parts of an app without auth is more immediately useful than having a way to disable that functionality.

Reply:
I just think that admins are going to want to put a limitation on this field. It's a big hole in security if teams can purposely or accidentally disable security per route and will want a way to lock this field down to specific roles in the cluster. Just my 2 cents.